notes/12-guarding-routes.md

```diff
@@ -0,0 +1,195 @@
+
+Import the action decorator at the top of the file:
+
+```js
+import { action } from '@ember/object';
+```
+
+Then add a property, `isAuthenticated` that will be used to store if the session is in an authenticated state, and an action method, `logout`, for invalidating the authenticated session and signing off the user.
+
+```js
+  get isAuthenticated() {
+    return !!this.currentUserId;
+  }
+
+  @action
+  logout() {
+    window.localStorage.removeItem(AUTH_KEY);
+    this.router.transitionTo('login');
+  }
+```
+
+In the login route, present at `app/routes/login.js`, add a `beforeModel` hook to perform session validations and handle the UI flow accordingly.
+
+In the login route, inject the imported `auth` service and add the `beforeModel` hook to perform session validation.
+
+```js
+   /**
+   * @type {AuthService}
+   */
+  @service auth;
+
+  async beforeModel(transition) {
+    await super.beforeModel(transition);
+    if (this.auth.isAuthenticated) {
+      this.transitionTo('teams');
+    }
+  }
+```
+
+And, in the same route file(`login` route), import the `auth` service, since its not already available(injected).
+
+```js
+import { inject as service } from '@ember/service';
+import AuthService from 'shlack/services/auth';
+```
+
+Now let's move on to `teams` route.
+Similar to `login` route, handle session validation in the `beforeModel` hook of the `teams` route after injecting the auth service into the route. There is a key difference to note here - the validation logic is flipped, that is the redirection will happen only when an authenticated session _does not exists_
+
+In the `teams` route, inject the imported `auth` service and add the `beforeModel` hook to perform session validation.
+
+```js
+// app/routes/teams.js
+
+   /**
+   * @type {AuthService}
+   */
+  @service auth;
+
+  async beforeModel(transition) {
+    await super.beforeModel(transition);
+    if (!this.auth.isAuthenticated) {
+      this.transitionTo('login');
+    }
+  }
+```
+
+And, in the same route file(`teams` route), import the `auth` service, since its not already available(injected).
+
+```js
+// app/routes/teams.js
+import { inject as service } from '@ember/service';
+import AuthService from 'shlack/services/auth';
+```
+
+## Changes in Templates
+
+Now that we have the javascript implementation in place, lets add the UI for the same.
+In `app/templates/components/team-sidebar.hbs`, replace the `LinkTo` component with a plain old HTML `button` element with an onclick handler that will trigger the `logout` action that you defined in `app/services/auth.js`.
+
+```diff
+-    <LinkTo @route='login' {{! destination route }}
+-            @tagName="button" {{! use <button> instead of <a> }}
+-            class="text-white rounded bg-grey-dark hover:bg-red-darker p-2 team-sidebar__logout-button" {{! HTML classes}}
++    <button {{on "click" this.auth.logout}} class="text-white rounded bg-grey-dark hover:bg-red-darker p-2 team-sidebar__logout-button"
```

Comment:
I think I'll adjust this solution so the `class=` is on its own line. This will make your diff a bit simpler, since that same line is present in both the `<button>` and `<LinkTo>`. Let's update this diff accordingly 

```diff
-    <LinkTo @route='login' {{! destination route }}
-            @tagName="button" {{! use <button> instead of <a> }}
-			 {{! HTML classes}}
             class="text-white rounded bg-grey-dark hover:bg-red-darker p-2 team-sidebar__logout-button"
+    <button {{on "click" this.auth.logout}}
			 class="text-white rounded bg-grey-dark hover:bg-red-darker p-2 team-sidebar__logout-button"
     >
```

Reply:
How do you recommend we move forward with this one ? Should I wait till code change is made to exercise solution ?